### PR TITLE
refactor: Refactor error variants and usage of them

### PR DESCRIPTION
**Pull request by ozgrakkurt**
closes https://github.com/semiotic-ai/timeline-aggregation-protocol/issues/73.

I found issues mostly about `InvalidCheckError` which was used as a generic error in the aggregator crate. I think I got all the cases.

### CODE REVIEW

**Review by ColePBryan — COMMENTED** [review]
Looks good to me. @aasseman can you give it a quick review as well before we merge since it mostly is hitting the aggregator.

**Review by aasseman — CHANGES_REQUESTED** [review]
Nice, just a few little things

**Comment by aasseman** [comment]
Also, are we OK with bundling all errors into a [JSON-RPC `-32002` error](https://github.com/semiotic-ai/timeline-aggregation-protocol/blob/341d0ab4232de7d5ccd366c201fd1b9798c81b54/tap_aggregator/src/error_codes.rs#L21) (current situation), or should we do something more fine-grained?
I'm fine with the current state of affairs, but maybe I'm wrong :shrug:

File: tap_aggregator/src/aggregator.rs [hunk 1]
```diff
@@ -36,22 +36,19 @@ pub async fn check_and_aggregate_receipts(
...
     // Check that the rav has the correct allocation id
     if let Some(previous_rav) = &previous_rav {
-        if previous_rav.message.allocation_id != allocation_id {
-            return Err(tap_core::Error::InvalidCheckError {
-                check_string: "Previous rav allocation id does not match receipts".into(),
+        let prev_id = previous_rav.message.allocation_id;
+        if prev_id != allocation_id {
+            return Err(tap_core::Error::RavAllocationIdMismatch {
+                prev_id: hex::encode(prev_id.as_bytes()),
+                new_id: hex::encode(allocation_id.as_bytes()),
```
[the remark this answers is not in the document]
Comment (ozgrakkurt) [reply tap_aggregator/src/aggregator.rs:51]: how can I do something like this for `signature` ?

**Comment by ozgrakkurt** [comment]
> Also, are we OK with bundling all errors into a [JSON-RPC `-32002` error](https://github.com/semiotic-ai/timeline-aggregation-protocol/blob/341d0ab4232de7d5ccd366c201fd1b9798c81b54/tap_aggregator/src/error_codes.rs#L21) (current situation), or should we do something more fine-grained? I'm fine with the current state of affairs, but maybe I'm wrong 🤷

should be different issue I think

File: tap_core/src/error.rs [hunk 2]
```diff
@@ -46,9 +46,9 @@ pub enum Error {
     #[error("Duplicate receipt signature: {0}")]
     DuplicateReceiptSignature(String),
     #[error(
-        "Receipt timestamp ({receipt_ts}) is greater than the previous RAV timestamp ({rav_ts})"
+        "Receipt timestamp ({receipt_ts}) is less or equal then previous rav timestamp ({rav_ts})"
```
Comment (aasseman) [inline tap_core/src/error.rs:49]: typo
```suggestion
        "Receipt timestamp ({receipt_ts}) is less or equal than previous rav timestamp ({rav_ts})"
```

File: tap_aggregator/src/aggregator.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ozgrakkurt) [reply tap_aggregator/src/aggregator.rs]: done

File: tap_aggregator/src/aggregator.rs [hunk 3]
```diff
@@ -103,9 +96,9 @@ fn check_receipt_timestamps(
         for receipt in receipts.iter() {
             let receipt = &receipt.message;
             if previous_rav.message.timestamp_ns > receipt.timestamp_ns {
-                return Err(tap_core::Error::InvalidCheckError {
-                    check_string: "Receipt timestamp is less or equal then previous rav timestamp"
-                        .into(),
+                return Err(tap_core::Error::ReceiptTimestampGreaterThanRav {
```
[the remark this answers is not in the document]
Comment (ozgrakkurt) [reply tap_aggregator/src/aggregator.rs]: done

**Review by aasseman — CHANGES_REQUESTED** [review]

**Comment by aasseman** [comment]
> [Merge branch 'main' into error-variant-refactor](https://github.com/semiotic-ai/timeline-aggregation-protocol/pull/112/commits/b17cb207773ba67054a71828c6cd7e16d762953c)

Could you rebase instead?

**Comment by ozgrakkurt** [comment]
> > [Merge branch 'main' into error-variant-refactor](https://github.com/semiotic-ai/timeline-aggregation-protocol/pull/112/commits/b17cb207773ba67054a71828c6cd7e16d762953c)
> 
> Could you rebase instead?

Actually wanted to just squash at the end into one commit but commit message checker is not gonna let me do it I think

File: tap_core/src/error.rs [hunk 4]
```diff
@@ -39,6 +39,16 @@ pub enum Error {
...
     #[error("Failed to produce rav request, no valid receipts")]
     NoValidReceiptsForRAVRequest,
+    #[error("Previous RAV allocation id ({prev_id}) doesn't match the allocation id from the new receipt ({new_id}).")]
+    RavAllocationIdMismatch { prev_id: String, new_id: String },
+    #[error("All receipts should have the same allocation id, but they don't")]
+    RavAllocationIdNotUniform,
+    #[error("Duplicate receipt signature: {0}")]
+    DuplicateReceiptSignature(String),
+    #[error(
+        "Receipt timestamp ({receipt_ts}) is less or equal than previous rav timestamp ({rav_ts})"
```
Comment (aasseman) [inline tap_core/src/error.rs:49]: `less or equal` does not correspond to the code (`strictly less`), BUT, turns out the code is wrong! So perhaps let's fix that line of code in another PR and merge it, then update this branch?

**Review by aasseman — CHANGES_REQUESTED** [review]

File: tap_core/src/error.rs (hunk 4 above)
Comment (aasseman) [reply tap_core/src/error.rs:49]: #115

Comment (aasseman) [reply tap_core/src/error.rs:49]: #115 is now fixed on the main branch. Could you update to latest main?

Comment (ozgrakkurt) [reply tap_core/src/error.rs:49]: done !

**Comment by ozgrakkurt** [comment]
@aasseman can you re-check? should be ok to merge

**Review by ColePBryan — APPROVED** [review]

**Review by aasseman — APPROVED** [review]

**Pull request merged**